ICSharpCode.Decompiler/IL/Transforms/InterpolatedStringTransform.cs

```diff
@@ -0,0 +1,149 @@
+		{
+			if (!context.Settings.StringInterpolation)
+				return;
+			int interpolationStart = pos;
+			int interpolationEnd;
+			Call insertionPoint;
+			// stloc v(newobj DefaultInterpolatedStringHandler..ctor(ldc.i4 literalLength, ldc.i4 formattedCount))
+			if (block.Instructions[pos] is StLoc
+				{
+					Variable: ILVariable { Kind: VariableKind.Local } v,
+					Value: NewObj { Arguments: { Count: 2 } } newObj
+				} stloc
+				&& v.Type.IsKnownType(KnownTypeCode.DefaultInterpolatedStringHandler)
+				&& newObj.Method.DeclaringType.IsKnownType(KnownTypeCode.DefaultInterpolatedStringHandler)
+				&& newObj.Arguments[0].MatchLdcI4(out int literalLength)
+				&& newObj.Arguments[1].MatchLdcI4(out int formattedCount))
+			{
+				// { call MethodName(ldloca v, ...) }
+				do
+				{
+					pos++;
+				}
+				while (IsKnownCall(block, pos, v));
+				interpolationEnd = pos;
+				// ... call ToStringAndClear(ldloca v) ...
+				if (!FindToStringAndClear(block, pos, v, out insertionPoint))
+				{
+					return;
+				}
+				if (!(v.StoreCount == 1 && v.AddressCount == interpolationEnd - interpolationStart))
+				{
+					return;
+				}
+			}
+			else
+			{
+				return;
+			}
+			context.Step($"Transform DefaultInterpolatedStringHandler {v.Name}", stloc);
+			var newVariable = v.Function.RegisterVariable(VariableKind.InitializerTarget, v.Type);
+			var replacement = new Block(BlockKind.InterpolatedString);
+			var initializer = stloc.Value;
+			stloc.Value = replacement;
+			replacement.Instructions.Add(new StLoc(newVariable, initializer));
+			for (int i = interpolationStart + 1; i < interpolationEnd; i++)
+			{
+				var inst = block.Instructions[i];
+				((LdLoca)((Call)inst).Arguments[0]).Variable = newVariable;
+				replacement.Instructions.Add(inst);
+			}
+			var callToStringAndClear = insertionPoint;
+			insertionPoint.ReplaceWith(replacement);
+			((LdLoca)callToStringAndClear.Arguments[0]).Variable = newVariable;
+			replacement.FinalInstruction = callToStringAndClear;
+			block.Instructions.RemoveRange(interpolationStart, interpolationEnd - interpolationStart);
+		}
+
+		private bool IsKnownCall(Block block, int pos, ILVariable v)
+		{
+			if (pos >= block.Instructions.Count - 1)
+				return false;
+			if (!(block.Instructions[pos] is Call call))
+				return false;
+			if (!(call.Arguments.Count > 1))
+				return false;
+			if (!call.Arguments[0].MatchLdLoca(v))
+				return false;
+			if (call.Method.IsStatic)
+				return false;
+			if (!call.Method.DeclaringType.IsKnownType(KnownTypeCode.DefaultInterpolatedStringHandler))
+				return false;
+			switch (call.Method.Name)
+			{
+				case "AppendLiteral" when call.Arguments.Count == 2 && call.Arguments[1] is LdStr:
+					break;
+				case "AppendFormatted" when call.Arguments.Count <= 4:
+					break;
+				default:
+					return false;
+			}
+			return true;
+		}
+
+		private bool FindToStringAndClear(Block block, int pos, ILVariable v, out Call insertionPoint)
```

Comment:
This can lead to re-ordering of side-effects. This needs to use `ILInlining.FindLoadInNext` as other similar transforms do